src/query/storage/m3/dynamic_cluster.go

```diff
@@ -88,10 +96,397 @@ func NewDynamicClusterOptions() DynamicClusterOptions {
+	for _, agg := range aggOpts.Aggregations() {
+		var (
+			clusterNamespace ClusterNamespace
+			err              error
+		)
+		if agg.Aggregated {
+			clusterNamespace, err = newAggregatedClusterNamespace(AggregatedClusterNamespaceDefinition{
+				NamespaceID: md.ID(),
+				Session:     clusterCfg.session,
+				Retention:   retOpts.RetentionPeriod(),
+				Resolution:  agg.Attributes.Resolution,
+				Downsample: &ClusterNamespaceDownsampleOptions{
+					All: agg.Attributes.DownsampleOptions.All,
+				},
+			})
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			clusterNamespace, err = newUnaggregatedClusterNamespace(UnaggregatedClusterNamespaceDefinition{
+				NamespaceID: md.ID(),
+				Session:     clusterCfg.session,
+				Retention:   retOpts.RetentionPeriod(),
+			})
+			if err != nil {
+				return nil, err
+			}
+		}
+		clusterNamespaces = append(clusterNamespaces, clusterNamespace)
+	}
+
+	return clusterNamespaces, nil
+}
+
+func (d *dynamicCluster) updateClusterNamespacesWithLock() {
+	nsCount := 0
+	for _, nsMap := range d.namespacesByEtcdCluster {
+		for _, clusterNamespaces := range nsMap.metadataToClusterNamespaces {
+			nsCount += len(clusterNamespaces)
+		}
+	}
+
+	var (
+		newNamespaces            = make(ClusterNamespaces, 0, nsCount)
+		newAggregatedNamespaces  = make(map[RetentionResolution]ClusterNamespace)
+		newUnaggregatedNamespace ClusterNamespace
+	)
+
+	for _, nsMap := range d.namespacesByEtcdCluster {
+		for _, clusterNamespaces := range nsMap.metadataToClusterNamespaces {
+			for _, clusterNamespace := range clusterNamespaces {
+				attrs := clusterNamespace.Options().Attributes()
+				if attrs.MetricsType == storagemetadata.UnaggregatedMetricsType {
+					if newUnaggregatedNamespace != nil {
+						d.logger.Warn("more than one unaggregated namespace found. using most recently "+
+							"discovered unaggregated namespace",
+							zap.String("existing", newUnaggregatedNamespace.NamespaceID().String()),
+							zap.String("new", clusterNamespace.NamespaceID().String()))
+					}
+					newUnaggregatedNamespace = clusterNamespace
+				} else {
+					retRes := RetentionResolution{
+						Retention:  attrs.Retention,
+						Resolution: attrs.Resolution,
+					}
+					existing, ok := newAggregatedNamespaces[retRes]
+					if ok {
+						d.logger.Warn("more than one aggregated namespace found for retention and resolution. "+
+							"using most recently discovered aggregated namespace",
+							zap.String("retention", retRes.Retention.String()),
+							zap.String("resolution", retRes.Resolution.String()),
+							zap.String("existing", existing.NamespaceID().String()),
+							zap.String("new", clusterNamespace.NamespaceID().String()))
+					}
+					newAggregatedNamespaces[retRes] = clusterNamespace
+				}
+			}
+		}
+	}
+
+	newNamespaces = append(newNamespaces, newUnaggregatedNamespace)
+	for _, ns := range newAggregatedNamespaces {
+		newNamespaces = append(newNamespaces, ns)
+	}
```

Comment:
`newAggregatedNamespaces` is a map, so this will not always yield deterministic order; not sure if that is important here though; if not ignore comment

Reply:
It looks like it's fine from the call sites, but I'll sort it just to be safe.

Reply:
No need to sort if it's not necessary, it's also a bit difficult to decide on a metric for sorting (resolution vs name vs retention?) so if the logic didn't exist already, it should be fine